### PR TITLE
feature:Participant report rollback

### PR DESCRIPTION
**Pull request by ruqinhu**
### Ⅰ. Describe what this PR did
Participant report rollback

### CODE REVIEW

File: tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java [hunk 1]
```diff
@@ -91,6 +92,10 @@ private void completeTransactionAfterThrowing(TransactionInfo txInfo, GlobalTran
                 throw new TransactionalExecutor.ExecutionException(tx, txe,
                         TransactionalExecutor.Code.RollbackFailure, ex);
             }
+            //throw to Launcher
+            if (tx.getRole() == GlobalTransactionRole.Participant) {
+                throw new TransactionException(TransactionExceptionCode.ParticipantReportRollback);
```
Comment (zjinlei) [inline tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java]: please throw original business exception.

File: tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java [hunk 2]
```diff
@@ -91,6 +92,10 @@ private void completeTransactionAfterThrowing(TransactionInfo txInfo, GlobalTran
                 throw new TransactionalExecutor.ExecutionException(tx, txe,
```
Comment (zjinlei) [inline tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java]: If participant rollback failed, What will happen?

File: tm/src/test/java/io/seata/tm/api/TransactionTemplateTest.java [hunk 3]
```diff
@@ -15,19 +15,20 @@
...
+import io.seata.core.exception.TransactionExceptionCode;
 import io.seata.core.model.GlobalStatus;
 import io.seata.core.model.TransactionManager;
 import io.seata.tm.TransactionManagerHolder;
-import io.seata.tm.api.transaction.NoRollbackRule;
-import io.seata.tm.api.transaction.RollbackRule;
-import io.seata.tm.api.transaction.TransactionHook;
-import io.seata.tm.api.transaction.TransactionHookManager;
-import io.seata.tm.api.transaction.TransactionInfo;
+import io.seata.tm.api.transaction.*;
```
Comment (zjinlei) [inline tm/src/test/java/io/seata/tm/api/TransactionTemplateTest.java]: not use *

Comment (ruqinhu) [reply tm/src/test/java/io/seata/tm/api/TransactionTemplateTest.java]: changed

Comment (ruqinhu) [reply tm/src/test/java/io/seata/tm/api/TransactionTemplateTest.java]: > not use *

changed

File: tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java (hunk 1 above)
Comment (ruqinhu) [reply tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java]: > please throw original business exception.

changed

Comment (zjinlei) [reply tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java]: original business exception is ex.
if rollbackDone, we need throw ex.

**Comment by zjinlei** [comment]
The first version added participantRollBackCode is closer to my idea; However, it needs to be expanded to handle 2 cases, Participants roll back success and failure.
TransactionExceptionCode.ParticipantRollbackDone and TransactionExceptionCode.ParticipantRollbackFailure is recommended to replace TransactionExceptionCode.ParticipantReportRollback.

File: tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java [hunk 4]
```diff
@@ -111,10 +117,28 @@ private void commitTransaction(GlobalTransaction tx) throws TransactionalExecuto
...
+        if (tx.getRole() == GlobalTransactionRole.Launcher) {
+            if (ex instanceof TransactionException) {
+                throw new TransactionalExecutor.ExecutionException(tx, TransactionalExecutor.Code.RollbackDone, ex.getCause());
+            } else {
+                throw new TransactionalExecutor.ExecutionException(tx, TransactionalExecutor.Code.RollbackDone, ex);
+            }
+        } else if (tx.getRole() == GlobalTransactionRole.Participant) {
+            if (ex instanceof TransactionException) {
+                throw new TransactionalExecutor.ExecutionException(tx, TransactionalExecutor.Code.RollbackDone,
+                        new TransactionException(TransactionExceptionCode.ParticipantRollbackDone, ex.getCause()));
```
Comment (zjinlei) [inline tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java]: ex.getCause()==ex
I think that only the code is different here, you can write a get code function, and completeTransactionAfterThrowing is also.

File: core/src/main/java/io/seata/core/exception/TransactionExceptionCode.java [hunk 5]
```diff
@@ -118,7 +118,17 @@
...
+
+    /**
+     * Participant rollback done
+     */
+    ParticipantRollbackDone,
+
+    /**
+     * Participant rollback failure
+     */
+    ParticipantRollbackFailure
```
Comment (zjinlei) [inline core/src/main/java/io/seata/core/exception/TransactionExceptionCode.java]: please move them to TransactionalExecutor.Code, then implement them in GlobalTransactionalInterceptor.handleGlobalTransaction.

Comment (ruqinhu) [reply core/src/main/java/io/seata/core/exception/TransactionExceptionCode.java]: > please move them to TransactionalExecutor.Code, then implement them in GlobalTransactionalInterceptor.handleGlobalTransaction.

ok

File: tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java (hunk 4 above)
Comment (ruqinhu) [reply tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java]: > ex.getCause()==ex
When using the construction method，ex.geCause() == originalException

File: spring/src/main/java/io/seata/spring/annotation/GlobalTransactionalInterceptor.java [hunk 6]
```diff
@@ -151,9 +153,12 @@ public TransactionInfo getTransactionInfo() {
                 case RollbackFailure:
                     failureHandler.onRollbackFailure(e.getTransaction(), e.getCause());
                     throw e.getCause();
+                case ParticipantRollbackDone:
+                    throw new TransactionException(TransactionExceptionCode.ParticipantReportedRollback,e.getOriginalException());
+                case ParticipantRollbackFailure:
+                    throw e.getCause();
```
Comment (zjinlei) [inline spring/src/main/java/io/seata/spring/annotation/GlobalTransactionalInterceptor.java]: A->B->C->D->E
if E throw e.getCause(), D, B, C, and A may all request a rollback with the TC. 
It is recommended that the participant fails to roll back, and the upstream link no longer requests the tc rollback.

File: spring/src/main/java/io/seata/spring/annotation/GlobalTransactionalInterceptor.java [hunk 7]
```diff
@@ -42,6 +39,11 @@
 import org.springframework.core.BridgeMethodResolver;
 import org.springframework.util.ClassUtils;
 
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import java.util.LinkedHashSet;
+import java.util.Set;
```
Comment (zjinlei) [inline spring/src/main/java/io/seata/spring/annotation/GlobalTransactionalInterceptor.java]: please restore, jdk import first.

**Comment by zjinlei** [comment]
wait for https://github.com/seata/seata/pull/2333 to merge

File: tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java [hunk 8]
```diff
@@ -259,4 +263,36 @@ private void cleanUp() {
...
+            return ex;
+        }
+    }
+
+    private TransactionalExecutor.Code getRollbackCode(GlobalTransaction tx) throws TransactionException {
+        if (tx.getLocalStatus() == GlobalStatus.RollbackRetrying) {
+            if (tx.getRole() == GlobalTransactionRole.Launcher) {
+                return TransactionalExecutor.Code.RollbackRetrying;
+            } else {
+                return TransactionalExecutor.Code.ParticipantRollbackDone;
```
Comment (zjinlei) [inline tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java:279]: This is the scenario where the rollback fails. You should not return to ParticipantRollbackDone.

File: tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java [hunk 9]
```diff
@@ -130,17 +131,17 @@ public boolean existingTransaction() {
...
-
-
     private void completeTransactionAfterThrowing(TransactionInfo txInfo, GlobalTransaction tx, Throwable ex) throws TransactionalExecutor.ExecutionException {
         //roll back
         if (txInfo != null && txInfo.rollbackOn(ex)) {
             try {
-                rollbackTransaction(tx, ex);
+                if (tx.getRole() == GlobalTransactionRole.Launcher || txInfo.getParticipantReportRollback()) {
+                    rollbackTransaction(tx, ex);
+                }
```
Comment (zjinlei) [inline tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java:140]: Changed the original logic, which will cause unParticipantReportRollback's participants not to take the TransactionHook.
Consider processing in rollbackTransaction-> tx.rollback.

Comment (ruqinhu) [reply tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java]: > Changed the original logic, which will cause unParticipantReportRollback's participants not to take the TransactionHook.
> Consider processing in rollbackTransaction-> tx.rollback.

changed

File: tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java (hunk 8 above)
Comment (ruqinhu) [reply tm/src/main/java/io/seata/tm/api/TransactionalTemplate.java]: > This is the scenario where the rollback fails. You should not return to ParticipantRollbackDone.

changed

**Pull request closed without merging**